File: prompt-ver-core/src/domain/diff.py
from dataclasses import dataclass
from difflib import SequenceMatcher
from enum import Enum
from uuid import UUID

from src.domain.prompt_version import PromptVersion


class DiffLineKind(str, Enum):
    ADDED = "ADDED"
    REMOVED = "REMOVED"
    EQUAL = "EQUAL"


@dataclass(frozen=True)
class DiffLine:
    kind: DiffLineKind
    content: str
    line_number_before: int | None
    line_number_after: int | None


@dataclass(frozen=True)
class VersionDiff:
    prompt_id: UUID
    version_id_before: UUID
    version_id_after: UUID
    version_number_before: int
    version_number_after: int
    lines: tuple[DiffLine, ...]

    @property
    def has_changes(self) -> bool:
        return any(line.kind != DiffLineKind.EQUAL for line in self.lines)


class VersionDiffer:
    def compare(self, before: PromptVersion, after: PromptVersion) -> VersionDiff:
        before_lines = before.content.value.splitlines()
        after_lines = after.content.value.splitlines()

        matcher = SequenceMatcher(None, before_lines, after_lines, autojunk=False)
        diff_lines: list[DiffLine] = []

        for tag, i1, i2, j1, j2 in matcher.get_opcodes():
            if tag == "equal":
                for k, line in enumerate(before_lines[i1:i2]):
                    diff_lines.append(DiffLine(
                        kind=DiffLineKind.EQUAL,
                        content=line,
                        line_number_before=i1 + k + 1,
                        line_number_after=j1 + k + 1,
                    ))
            elif tag == "replace":
                for k, line in enumerate(before_lines[i1:i2]):
                    diff_lines.append(DiffLine(
                        kind=DiffLineKind.REMOVED,
                        content=line,
                        line_number_before=i1 + k + 1,
                        line_number_after=None,
                    ))
                for k, line in enumerate(after_lines[j1:j2]):
                    diff_lines.append(DiffLine(
                        kind=DiffLineKind.ADDED,
                        content=line,
                        line_number_before=None,
                        line_number_after=j1 + k + 1,
                    ))
            elif tag == "delete":
                for k, line in enumerate(before_lines[i1:i2]):
                    diff_lines.append(DiffLine(
                        kind=DiffLineKind.REMOVED,
                        content=line,
                        line_number_before=i1 + k + 1,
                        line_number_after=None,
                    ))
            elif tag == "insert":
                for k, line in enumerate(after_lines[j1:j2]):
                    diff_lines.append(DiffLine(
                        kind=DiffLineKind.ADDED,
                        content=line,
                        line_number_before=None,
                        line_number_after=j1 + k + 1,
                    ))

        return VersionDiff(
            prompt_id=before.prompt_id,
            version_id_before=before.id,
            version_id_after=after.id,
            version_number_before=before.version_number.value,
            version_number_after=after.version_number.value,
            lines=tuple(diff_lines),
        )

File: prompt-ver-front/src/utils/diff.ts
import type { DiffLineResponse, VersionDiffResponse } from '../types';

function buildLCS(a: string[], b: string[]): number[][] {
  const m = a.length;
  const n = b.length;
  const dp: number[][] = Array.from({ length: m + 1 }, () => new Array(n + 1).fill(0));
  for (let i = 1; i <= m; i++) {
    for (let j = 1; j <= n; j++) {
      dp[i][j] = a[i - 1] === b[j - 1] ? dp[i - 1][j - 1] + 1 : Math.max(dp[i - 1][j], dp[i][j - 1]);
    }
  }
  return dp;
}

function backtrack(dp: number[][], a: string[], b: string[], i: number, j: number): DiffLineResponse[] {
  if (i === 0 && j === 0) return [];
  if (i === 0) {
    return [
      ...backtrack(dp, a, b, i, j - 1),
      { kind: 'insert', content: b[j - 1], line_number_before: null, line_number_after: j },
    ];
  }
  if (j === 0) {
    return [
      ...backtrack(dp, a, b, i - 1, j),
      { kind: 'delete', content: a[i - 1], line_number_before: i, line_number_after: null },
    ];
  }
  if (a[i - 1] === b[j - 1]) {
    return [
      ...backtrack(dp, a, b, i - 1, j - 1),
      { kind: 'equal', content: a[i - 1], line_number_before: i, line_number_after: j },
    ];
  }
  if (dp[i - 1][j] >= dp[i][j - 1]) {
    return [
      ...backtrack(dp, a, b, i - 1, j),
      { kind: 'delete', content: a[i - 1], line_number_before: i, line_number_after: null },
    ];
  }
  return [
    ...backtrack(dp, a, b, i, j - 1),
    { kind: 'insert', content: b[j - 1], line_number_before: null, line_number_after: j },
  ];
}

export function computeDiff(
  promptId: string,
  versionIdBefore: string,
  versionIdAfter: string,
  versionNumberBefore: number,
  versionNumberAfter: number,
  contentBefore: string,
  contentAfter: string
): VersionDiffResponse {
  const aLines = contentBefore.split('\n');
  const bLines = contentAfter.split('\n');
  const dp = buildLCS(aLines, bLines);
  const lines = backtrack(dp, aLines, bLines, aLines.length, bLines.length);
  const hasChanges = lines.some((l) => l.kind !== 'equal');

  return {
    prompt_id: promptId,
    version_id_before: versionIdBefore,
    version_id_after: versionIdAfter,
    version_number_before: versionNumberBefore,
    version_number_after: versionNumberAfter,
    lines,
    has_changes: hasChanges,
  };
}
